### PR TITLE
ipc: multidai abi

### PR DESCRIPTION
**Pull request by slawblauciak**
Introduces configuration IPC ABI for the MultiDAI

MultiDAI is a new logical DAI type responsible for allocation of multiple physical child DAI and transparently managing them, routing the audio data.

Signed-off-by: Slawomir Blauciak <slawomir.blauciak@linux.intel.com>

### CODE REVIEW

File: src/include/ipc/dai-intel.h [hunk 1]
```diff
@@ -196,4 +197,27 @@ struct sof_ipc_dai_dmic_params {
...
+struct sof_ipc_dai_multidai_params {
+	struct sof_ipc_hdr hdr;
+	uint32_t num_dais;
+
+	uint32_t reserved[6];
+
+	struct sof_ipc_stream_map stream_map;
+	struct sof_ipc_dai_multidai_ext ext[0];
+} __attribute__((packed));
+
```
Comment (lgirdwood) [inline src/include/ipc/dai-intel.h]: Why would we not want to make this generic in dai.h ? 
I would prefer we have a parent params structure that would contain an array of sof_ipc_dai_config. This would allow reuse a lot of existing logic in the driver/fw and also mean we could pack in different types together.

File: src/include/ipc/dai-intel.h [hunk 2]
```diff
@@ -196,4 +197,27 @@ struct sof_ipc_dai_dmic_params {
...
+	};
+} __attribute__((packed));
+
+struct sof_ipc_dai_multidai_params {
+	struct sof_ipc_hdr hdr;
+	uint32_t num_dais;
+
+	uint32_t reserved[6];
+
+	struct sof_ipc_stream_map stream_map;
```
Comment (lgirdwood) [inline src/include/ipc/dai-intel.h]: We probably need a stream_map per dai config (since TDM modes can remap channels).

File: src/include/ipc/dai-intel.h (hunk 1 above)
Comment (slawblauciak) [reply src/include/ipc/dai-intel.h]: I made it multidai specific to avoid major ABI changes.

File: src/include/ipc/dai-intel.h (hunk 2 above)
Comment (slawblauciak) [reply src/include/ipc/dai-intel.h]: Can you please elaborate what you mean by TDM mode channel remapping? This might be something entirely different,  so it might have to land in the SSP-specific IPC.

Comment (lgirdwood) [reply src/include/ipc/dai-intel.h]: In TDM mode, SSP can support upto 8 channels and these can be remapped to SSP slots differently from L/R mode. i.e. consider DAI config 0 is L/R mode stereo (slot 0,1), whilst DAI config 1 is 4 channel mode with L/R mapped to (slot 2,3).

File: src/include/ipc/dai-intel.h (hunk 1 above)
Comment (lgirdwood) [reply src/include/ipc/dai-intel.h]: Should juyst be a MINOR ABI change. We need to create a new struct that can contain sof_ipc_dai_config

Comment (slawblauciak) [reply src/include/ipc/dai-intel.h]: Oh, sorry, I misunderstood what you mean. Now that you mentioned it, yes, I agree.

File: src/include/ipc/dai-intel.h (hunk 2 above)
Comment (slawblauciak) [reply src/include/ipc/dai-intel.h]: Yeah, that's SSP-specific, unrelated to the purpose of the map contained in multidai params.
It's for routing the stream channels to individual physical child DAIs.

File: src/include/ipc/dai-intel.h (hunk 1 above)
Comment (slawblauciak) [reply src/include/ipc/dai-intel.h]: Having thought about this, I don't thing there's a sane way to do that. Such a mechanic introduces circular dependency between sof_ipc_dai_multidai_params and sof_ipc_dai_config.

Comment (lgirdwood) [reply src/include/ipc/dai-intel.h]: How will we have a circular dependency ? They should both be defined in the same header ?

Comment (slawblauciak) [reply src/include/ipc/dai-intel.h]: They both would have an instance (or more) of each other:
sof_ipc_dai_multidai_params having a flex array of sof_ipc_dai_configs and sof_ipc_dai_config having sof_ipc_dai_multidai_params in the union

File: src/include/ipc/dai-intel.h [hunk 3]
```diff
@@ -17,6 +17,7 @@
 #define __IPC_DAI_INTEL_H__
 
 #include <ipc/header.h>
```
Comment (dbaluta) [inline src/include/ipc/dai-intel.h]: @slawblauciak please add in the commit message more information about multi dai. 

You have added some information in the PR description:

"MultiDAI is a new logical DAI type responsible for allocation of multiple physical child DAI and transparently managing them, routing the audio data."

This must be present in the commit message because PR description will most likely be buried in the dark corners of github.

Comment (dbaluta) [reply src/include/ipc/dai-intel.h]: @slawblauciak also use a verb in the commit message. Like: "Introduce multi DAI abi".

File: src/include/ipc/dai-intel.h (hunk 1 above)
Comment (lgirdwood) [reply src/include/ipc/dai-intel.h]: oh, I would not expect sof_ipc_dai_config to contain the multidai type.
I would rename sof_ipc_dai_multidai_params to sof_ipc_dai_multidai_config which would be a new toplevel IPC and can contain an array of  sof_ipc_dai_config

Comment (slawblauciak) [reply src/include/ipc/dai-intel.h]: Ha, thinking out of the box. Didn't even cross my mind. That is a possibility indeed.
@plbossart @bardliao what do you think?

Comment (lgirdwood) [reply src/include/ipc/dai-intel.h]: This is MINOR ABI update and will require a mechanical kernel update too (i.e. if we detect > 1 DAI config then use new IPC)

Comment (slawblauciak) [reply src/include/ipc/dai-intel.h]: Changes applied.

File: src/include/ipc/dai.h [hunk 4]
```diff
@@ -87,4 +89,18 @@ struct sof_ipc_dai_config {
...
+	uint16_t format;	/**< SOF_DAI_FMT_ */
+	uint16_t reserved16;	/**< alignment */
+
+	uint32_t num_dais;
+	uint32_t reserved[5];
+
+	struct sof_ipc_stream_map stream_map;
+	struct sof_ipc_dai_config configs[0];
+} __attribute__((packed));
+
```
Comment (plbossart) [inline src/include/ipc/dai.h]: sorry, I don't understand the logic or rationale. And GitHub makes it really hard to follow the discussion between @slawblauciak and @lgirdwood since the initial commits are gone.

From my perspective, if the MULTIDAI is listed in the dai_type, then its configuration should be passed to the firmware in the same way as other DAIs. I don't see why we would have a new IPC, this is extra work for no good reason.

**Comment by bardliao** [comment]
@slawblauciak The concept is to put/link physical dai configs to the new virtual dal config, right? My question is what is format used for? Shouldn't we use the format of each physical dais?

File: src/include/ipc/dai.h [hunk 5]
```diff
@@ -61,6 +62,7 @@ enum sof_ipc_dai_type {
 	SOF_DAI_INTEL_ALH,		/**< Intel ALH */
 	SOF_DAI_IMX_SAI,                /**< i.MX SAI */
 	SOF_DAI_IMX_ESAI,               /**< i.MX ESAI */
+	SOF_DAI_MULTIDAI		/**< MultiDAI */
```
Comment (dbaluta) [inline src/include/ipc/dai.h]: comma after last enum.
```SOF_DAI_MULTIDAI,```

File: src/include/ipc/dai.h (hunk 4 above)
Comment (slawblauciak) [reply src/include/ipc/dai.h]: Yes, that was my initial proposal. If it's particularly inconvenient for you I can revert it.

**Comment by slawblauciak** [comment]
> @slawblauciak The concept is to put/link physical dai configs to the new virtual dal config, right? My question is what is format used for? Shouldn't we use the format of each physical dais?

If we have separate formats for each DAI, I suppose MultiDAI would be responsible for conversion?
Would we really have a use case for that?

**Comment by slawblauciak** [comment]
Changes applied. The code is now back to the initial proposal.

**Comment by bardliao** [comment]
> > @slawblauciak The concept is to put/link physical dai configs to the new virtual dal config, right? My question is what is format used for? Shouldn't we use the format of each physical dais?
> 
> If we have separate formats for each DAI, I suppose MultiDAI would be responsible for conversion?
> Would we really have a use case for that?

No, I don't think that we have different format for each DAI use case for now. However, I don't think we have to convert the format to a single format. To me, they are still two separate DAIs. We just want to enable them at the same time.

File: src/include/ipc/dai.h (hunk 4 above)
Comment (lgirdwood) [reply src/include/ipc/dai.h]: My understanding is that we would have a DAI, say SSP0 and be able to load multiple different HW configurations for the DAI in one IPC (with an index/ID for each config). This aligns to the ASoC topology ABI. It also allows loading all DAI configs in a single IPC.

Comment (plbossart) [reply src/include/ipc/dai.h]: @lgirdwood I think we are talking about different problems.

a) The discussion here is to define one 'multi-DAI' config IPC, so that we can enable synchronized transport on e.g. SSP0 and SSP4, or ALH1.3 and ALH 2.5 (the integer part is the link ID and the decimal the PDI number). So we send one compound configuration, and the firmware takes care of distributing the values at the low-level (since the kernel only interacts with a multi-dai).

n) What you describe in your comment is the ability to send multiple configurations for the same DAI (which could be extended for the multi-dai). for this we would need in addition a 'config_id' and a means to identify which config_id to apply. For example this could be the case for an SSP with support for 16kHz in BT HFP mode and 48kHz in A2DP mode. The selection of the config could be based on the_params, but it's not clear to me if the selection if done in firmware or at the kernel level.

We *may* combine the two topics and define a new structure that works for
- a)
- b)
- all of the above.

It's probably worth talking tomorrow morning on this, we're probably hitting the limits of off-line design

Comment (lgirdwood) [reply src/include/ipc/dai.h]: @plbossart ack, that's what I was thinking - the same structure should be usable for both (the encoded DAI ID in each struct can be used by FW to config both a) and b) even simultaneously with the same IPC)

Comment (plbossart) [reply src/include/ipc/dai.h]: ok, but then the definitions are not quite right

1. there should be an explicit number of configs
2. the num_dais and stream_map are only relevant for the b) case, so don't they belong in the configs[] field?

I think we should change to have a multi-config DAI IPC, that's backwards compatible with the existing one (i.e. the kernel would use the multi-config defintions even for a single config), and we should hide the 'virtual' dai as a regular DAI.

Comment (slawblauciak) [reply src/include/ipc/dai.h]: Sounds reasonable. If I understand correctly, we'd keep the dai config struct as it is, use one of the reserved fields to pass the number of configs and turn the union of dai-specific configs into a flex array.

Green light?

Comment (slawblauciak) [reply src/include/ipc/dai.h]: Hm, actually. How exactly are we supposed to have it backwards compatible?
It would be easy if there was a free reserved field at the beginning of the structure, where we'd have the number of configs. But we don't.

The only way I see here is actually duplicating the generic config fields and putting them in a flex-array but I honestly don't like this at all...

**Comment by slawblauciak** [comment]
Changes applied partially. Rest of the code will need to be changed to follow.
@plbossart @lgirdwood is this what you had in mind?

File: src/include/ipc/dai.h (hunk 4 above)
Comment (lgirdwood) [reply src/include/ipc/dai.h]: I guess so if I'm reading this correctly and it's backwards compatible.

**Comment by slawblauciak** [comment]
@lgirdwood just realized, the way I made it doesn't allow us to have multiple formats in a single config, since those are not part of the flex array. Is that okay?

**Comment by lgirdwood** [comment]
> @lgirdwood just realized, the way I made it doesn't allow us to have multiple formats in a single config, since those are not part of the flex array. Is that okay?

Sorry not, following, what do you mean by multiple formats ? Do you mean a single (SSP) DAI with multiple HW formats or do you mean multiple different DAIS with different formats ? @plbossart 

I would expect us to be able to send an array of DAI configs that can target multiple DAIs with multiple HW config formats/clocks/modes, although our initial use case is a single DAI with multiple configs (i.e. SSP0 with config 0 (L/R I2S stereo), SSP0 with config 1 (TDM mode, 4 channels)...

**Comment by plbossart** [comment]
> Sorry not, following, what do you mean by multiple formats ? Do you mean a single (SSP) DAI with multiple HW formats or do you mean multiple different DAIS with different formats ? @plbossart

I'd like support multiple DAIs of the same type but with the same list of formats each.
The case with DAIs of different types and different number of formats each is unmanageable.

**Comment by slawblauciak** [comment]
In order to have the ability to support multiple formats in a single IPC (the same way we support multiple configs in a single IPC) we'd need to have those formats also in a separate flex array. But then, we're literally duplicating the fields in a single struct. That makes the ABI far more confusing, because then it gets all kinds of blurry as to which field is to be actually used here.

File: src/audio/dai.c [hunk 6]
```diff
@@ -705,10 +705,10 @@ static int dai_config(struct comp_dev *dev, struct sof_ipc_dai_config *config)
 		return -EINVAL;
 	}
 
-	switch (config->type) {
+	switch (config->info.type) {
 	case SOF_DAI_INTEL_SSP:
 		/* set dma burst elems to slot number */
-		dd->config.burst_elems = config->ssp.tdm_slots;
+		dd->config.burst_elems = config->ssp[0].tdm_slots;
```
Comment (lgirdwood) [inline src/audio/dai.c]: rather than using 0, we should use a struct member (hard coded to 0 today), that way the final functional update becomes easier.

File: src/include/ipc/dai.h [hunk 7]
```diff
@@ -61,29 +62,53 @@ enum sof_ipc_dai_type {
...
+
+struct sof_ipc_multidai_params {
+	struct sof_ipc_cmd_hdr hdr;
+
+	uint32_t num_dais;
+	uint32_t reserved[5];
+
+	struct sof_ipc_stream_map stream_map;
+	struct sof_ipc_dai_info info[0];
+	struct sof_ipc_dai_base_params params[0];
```
Comment (lgirdwood) [inline src/include/ipc/dai.h]: This is difficult to index, best to do 
```
struct ipc_dai_blah {
    struct sof_ipc_dai_info info;
    struct sof_ipc_dai_base_params params;
};
```
then
```
struct sof_ipc_multidai_params {
...
    struct ipc_dai_blah dai_blah[0];
};
```
This way info and params can be addressed using regular array indexing (instead of manual calculations).

File: src/audio/dai.c (hunk 6 above)
Comment (slawblauciak) [reply src/audio/dai.c]: Not sure what you mean.
Are you talking about simply dereferencing 'ssp' instead of indexing it with 0?

Comment (slawblauciak) [reply src/audio/dai.c]: @lgirdwood sorry, mind giving me a hint here?

Comment (lgirdwood) [reply src/audio/dai.c]: Sorry, I mean
```
dd->dai_index = 0;
...
dd->config.burst_elems = config->ssp[dd->dai_index].tdm_slots;
```

Comment (slawblauciak) [reply src/audio/dai.c]: Ah, in the DAI data. Got it, cheers

Comment (slawblauciak) [reply src/audio/dai.c]: Hmm, on the other hand, this creates an inconsistency, because the drivers will still resort to indexing the 0.

One way I'd see here is to have the proposed dai_index (or rather config_index) inside of the DAI config IPC. However, this creates a potential issue if the data is not initialized to 0 in the IPC.

Comment (lgirdwood) [reply src/audio/dai.c]: @slawblauciak doesn't matter where you store the index today, as long as we have an index rather than hard coding to 0. The DAI config switching logic will no doubt place the index in the correct [place when it's implemented.

**Comment by slawblauciak** [comment]
Applied changes and rebased.

File: src/include/ipc/dai.h [hunk 8]
```diff
@@ -74,16 +74,37 @@ struct sof_ipc_dai_config {
...
+} __attribute__((packed));
+
+struct sof_ipc_multidai_params {
+	struct sof_ipc_cmd_hdr hdr;
+
+	uint32_t num_dais;
+	uint32_t reserved[5];
+
+	struct sof_ipc_stream_map stream_map;
+	struct sof_ipc_dai_base_params params[0];
```
Comment (plbossart) [inline src/include/ipc/dai.h]: I am a bit lost here.

The sof_ipc_dai_config already contains a 'sof_ipc_dai_base_params'.

So in the multi-dai case, is the baseline ignored?

Also in the size of this array determined by 'num_dais'?

File: src/include/ipc/dai.h [hunk 9]
```diff
@@ -74,16 +74,37 @@ struct sof_ipc_dai_config {
...
+} __attribute__((packed));
+
+/* general purpose DAI configuration */
+struct sof_ipc_dai_config {
+	struct sof_ipc_cmd_hdr hdr;
+	struct sof_ipc_dai_base_params params;
+
+	/* reserved for future use */
+	uint32_t reserved[3];
+	uint32_t num_configs;
```
Comment (plbossart) [inline src/include/ipc/dai.h]: are you positive that this change is compatible with the previous solution? I haven't checked the sizes and it's hard to see with the diff format.

Comment (slawblauciak) [reply src/include/ipc/dai.h]: Yeah. I ran tests on our side as well.

File: src/include/ipc/dai.h (hunk 8 above)
Comment (slawblauciak) [reply src/include/ipc/dai.h]: The base params inside of sof_ipc_dai_config is for the DAI you're directly communicating with.
The base params inside of the multidai is for the children of the multidai. And yes, the size is determined by num_dais.

To explain a bit more what's going on here:
The union of arrays contained within sof_ipc_dai_config contain structures, which don't convey basic DAI information (dai type and index). So those need to be passed separately. That's the reason behind the additional base params array.

Comment (plbossart) [reply src/include/ipc/dai.h]: let's take the example of a multi-dai SounDwire/ALH case.

The dai_config would contain
hdr
base_params (multi_DAI type + index)

So then the multidai array should contain the ALH information for each child DAI, right?

The problem is that I don't see anything ALH-related in the multi_dai_params...only a stream_map and a base_params (which can only mean individual DAI and type). Where would the ALH stream_id passed?

Comment (slawblauciak) [reply src/include/ipc/dai.h]: As I understood from the previous discussion, ALH related params would be in the flex array of the parent struct, which is in sof_ipc_dai_config.

To be specific: the first element of that flex array in this case would be the sof_ipc_dai_multidai_params. And directly after that, the very next members would be the children for multida, namely sof_ipc_dai_alh_params.

Comment (plbossart) [reply src/include/ipc/dai.h]: sorry, I don't get how this would work.
You don't have an array of unions, you have a union of arrays.

I don't think you can make the type of structure vary between array elements?

Comment (slawblauciak) [reply src/include/ipc/dai.h]: I'm pretty sure you can, since from the standpoint of the compiler, all union members have a size of 0 anyway. So, the union is purely cosmetic, serving as a template for the code. Now that I think about this, the union doesn't do anything at all here, might as well get rid of it.

Comment (slawblauciak) [reply src/include/ipc/dai.h]: Honestly, the more I think about this, the more I realize that there's no good way of actually having multiple configs for different DAI types in a single IPC.

- As you mentioned, having a union of flex arrays suggests we can only have 1 type, doing otherwise is indeed bad practice.
- Having a flex array of unions has padding issues I mentioned earlier.
- Having multiple flex arrays, without a union also doesn't smell right

I'd honestly abandon those efforts at this point and revert to the initial proposal: not touching the DAI config, simply adding a new DAI config member with its own union of flex arrays, limited to a single type.

Comment (q009) [reply src/include/ipc/dai.h]: @lgirdwood @plbossart can we come to an agreement on this one?

Comment (slawblauciak) [reply src/include/ipc/dai.h]: @lgirdwood @plbossart can we come to an agreement on this one?

Comment (lgirdwood) [reply src/include/ipc/dai.h]: @slawblauciak is the sticking point the unions or the ALH ? I defer to @plbossart for all ALH and Soundwire stuff.

Comment (q009) [reply src/include/ipc/dai.h]: It's the unions. I don't think it's sane to try to pack all the different configurations into a single IPC  inside of sof_ipc_dai_config. As I mentioned, I'd rather do this by the book and just do as in the initial proposal: a new union member in sof_ipc_dai_config, one that would have its own, similar union, except of flex arrays.

**Comment by jajanusz** [comment]
Is this PR active?

**Comment by slawblauciak** [comment]
Ah, nope, it is not.
